samples: nrf9160: add nRF Cloud REST Device Message sample (#6835)
nrfconnect/sdk-nrf: samples/nrf9160/nrf_cloud_rest_device_message/README.rst

.. _nrf_cloud_rest_device_message:

nRF9160: nRF Cloud REST Device Message
######################################

.. contents::
   :local:
   :depth: 2

The REST Device Message sample demonstrates how to use the `nRF Cloud REST API`_ to send a `Device Message <nRF Cloud Device Messages_>`_ using the ``SendDeviceMessage`` endpoint. The d2c topic is used in non-bulk mode.

Requirements
************

The sample supports the following development kits:

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading
   :rows: nrf9160dk_nrf9160_ns

The sample requires an `nRF Cloud`_ account.

.. note::
   This sample requires modem firmware v1.3.x or later.

Overview
********

> Pressing the button sends a button press message that you can view in the Terminal card of the `nRF Cloud`_ portal (check the Device-to-Cloud channel). The device must already be provisioned. If your device is not yet profisioned, you can do so over REST by enabling the :kconfig:`CONFIG_REST_DEVICE_MESSAGE_DO_JITP` option and following the instructions printed to UART upon boot.

Reviewer: ```suggestion
Pressing the button sends a button press message that you can view in the Terminal card of the `nRF Cloud`_ portal (check the Device-to-Cloud channel).
The device must already be provisioned.
If your device is not yet provisioned, you can do so over REST by enabling the :kconfig:`CONFIG_REST_DEVICE_MESSAGE_DO_JITP` option and following the instructions printed to UART upon boot.
```
Fixed typo (profisioned -> provisioned). Remember, every sentence on its own line.

Limitations
***********

This sample requires the network carrier to provide date and time to the modem.
Without a valid date and time, the modem cannot generate JWTs with an expiration time.
Note that REST connections are transient, unlike MQTT connections which are usually long-lived.

User interface
**************

Once provisioned and connected, each press of the configured button (configured using the :kconfig:`CONFIG_REST_DEVICE_MESSAGE_BUTTON_EVT_NUM` option) generates a device-to-cloud button press `Device Message <nRF Cloud Device Messages_>`_ over REST.

If you have the option :kconfig:`CONFIG_REST_DEVICE_MESSAGE_DO_JITP` enabled and you press the button when prompted at startup, it will request just-in-time provisioning (JITP) through REST with nRF Cloud.
This is useful when initially provisioning and associating a device on nRF Cloud.
You only need to do this once for each device.

The configured LTE LED (:kconfig:`CONFIG_REST_DEVICE_MESSAGE_LTE_LED_NUM`) is lit once an LTE connection is established and JWT tokens are ready to be generated.

> The configured Send LED (:kconfig:`CONFIG_REST_DEVICE_MESSAGE_SEND_LED_NUM`) is blinked every time a Device Message or JITP request is sent.

Reviewer: ```suggestion
The configured Send LED (:kconfig:`CONFIG_REST_DEVICE_MESSAGE_SEND_LED_NUM`) blinks every time a Device Message or JITP request is sent.
```

.. _nrf_cloud_rest_device_message:

nRF9160: nRF Cloud REST Device Message
######################################

.. contents::
   :local:
   :depth: 2

> The REST Device Message sample demonstrates how to use the `nRF Cloud REST API`_ to send a `Device Message <nRF Cloud Device Messages_>`_ using the ``SendDeviceMessage`` endpoint. The d2c topic is used in non-bulk mode.

Reviewer: The d2c topic might need some clarification. It just comes out of the blue.

Requirements
************

The sample supports the following development kits:

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading
   :rows: nrf9160dk_nrf9160_ns

The sample requires an `nRF Cloud`_ account.

.. note::
   This sample requires modem firmware v1.3.x or later.

Overview
********

> Pressing the button sends a button press message that you can view in the Terminal card of the `nRF Cloud`_ portal (check the Device-to-Cloud channel).

Reviewer: This makes the reader wonder which button and did they need to configure something first.
What does the Device-to-Cloud channel refer to?
Should we start from the prerequisites first? Something like:
"To start working with this sample, make sure your device is provisioned (connected to cloud). If not, enable the nnn Kconfig option and follow the instructions..."

The device must already be provisioned.
If your device is not yet provisioned, you can do so over REST by enabling the :kconfig:`CONFIG_REST_DEVICE_MESSAGE_DO_JITP` option and following the instructions printed to UART upon boot.

Limitations
***********

This sample requires the network carrier to provide date and time to the modem.

> Without a valid date and time, the modem cannot generate JWTs with an expiration time.

Reviewer: JWT comes suddenly in. Might be good to explain first why the JWT is needed.

Note that REST connections are transient, unlike MQTT connections which are usually long-lived.

User interface
**************

> 

Reviewer: You might want to consider describing the buttons and LEDs in this section like in https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/openthread/coap_client/README.html

> Once provisioned and connected, each press of the configured button (configured using the :kconfig:`CONFIG_REST_DEVICE_MESSAGE_BUTTON_EVT_NUM` option) generates a device-to-cloud button press `Device Message <nRF Cloud Device Messages_>`_ over REST.

Reviewer: ```suggestion
Once the device is provisioned and connected, each press of the configured button (configured using the :kconfig:`CONFIG_REST_DEVICE_MESSAGE_BUTTON_EVT_NUM` option) generates a device-to-cloud button press `Device Message <nRF Cloud Device Messages_>`_ over REST.
```

If you have the option :kconfig:`CONFIG_REST_DEVICE_MESSAGE_DO_JITP` enabled and you press the button when prompted at startup, it will request just-in-time provisioning (JITP) through REST with nRF Cloud.
This is useful when initially provisioning and associating a device on nRF Cloud.
You only need to do this once for each device.

The configured LTE LED (:kconfig:`CONFIG_REST_DEVICE_MESSAGE_LTE_LED_NUM`) is lit once an LTE connection is established and JWT tokens are ready to be generated.

> The configured Send LED (:kconfig:`CONFIG_REST_DEVICE_MESSAGE_SEND_LED_NUM`) is lit whenever a Device Message or JITP request is being sent.

Reviewer: ```suggestion
The configured Send LED (:kconfig:`CONFIG_REST_DEVICE_MESSAGE_SEND_LED_NUM`) is lit when a Device Message or JITP request is being sent.
```

CONFIG_REST_DEVICE_MESSAGE_LTE_LED_NUM - LTE LED number
   This configuration option defines which LED is used to indicate LTE connection success

.. kconfig:: CONFIG_REST_DEVICE_MESSAGE_SEND_LED_NUM

CONFIG_REST_DEVICE_MESSAGE_SEND_LED_NUM - Send LED number
   This configuration option defines which LED is used to indicate a REST request is being sent

.. kconfig:: CONFIG_REST_DEVICE_MESSAGE_BUTTON_EVT_NUM

CONFIG_REST_DEVICE_MESSAGE_BUTTON_EVT_NUM - Button number
   This configuration option defines the button to use for device interactions.

.. kconfig:: CONFIG_REST_DEVICE_MESSAGE_DO_JITP

CONFIG_REST_DEVICE_MESSAGE_DO_JITP - Enable prompt to perform JITP over REST
   This configuration option defines whether the application prompts the user for just-in-time provisioning on startup.

Building and running
********************

.. |sample path| replace:: :file:`samples/nrf9160/nrf_cloud_rest_device_message`

.. include:: /includes/build_and_run_nrf9160.txt

The configuration file for this sample is located in :file:`samples/nrf9160/nrf_cloud_rest_device_message`.
See :ref:`configure_application` for information on how to configure the parameters.

> Checking Device Messages over REST API
> **************************************

Reviewer: ```suggestion
Querying Device Messages over REST API
**************************************
```

.. _nrf_cloud_rest_device_message:

nRF9160: nRF Cloud REST Device Message
######################################

.. contents::
   :local:
   :depth: 2

The REST Device Message sample demonstrates how to use the `nRF Cloud REST API`_ to send `Device Messages <nRF Cloud Device Messages_>`_ using the ``SendDeviceMessage`` REST endpoint.

Requirements
************

The sample supports the following development kits:

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading
   :rows: nrf9160dk_nrf9160_ns

The sample requires an `nRF Cloud`_ account.

Your device must be provisioned with `nRF Cloud`_. If it is not, follow the instructions in `nrf_cloud_rest_device_message_provisioning`_.

.. note::
   This sample requires modem firmware v1.3.x or later.

Limitations
***********

The `nRF Cloud REST API`_ requires all requests to be authenticated with a JSON Web Token (JWT).

> See  `nRF Cloud Security`_ for more details.

Reviewer: Link target missing from the links.txt. Added one in my suggestion.

Generating valid JWTs requires the network carrier to provide date and time to the modem, so the sample must first connect to an LTE carrier and determine the current date and time before REST requests can be sent.

Note also that the `nRF Cloud REST API`_ is stateless.

> Devices may make REST requests at will.

Reviewer: Really? Isn't it the user of the sample that is making the REST calls (requests)? Is this sentence needed at all when you compare REST to MQTT?

Author: Just to be clear, you are suggesting that I remove: `Devices may make REST requests at will.`

I think that could work; I wanted to elaborate a bit on what I meant by stateless, but I guess the comparison to MQTT does make that clear to a reader

Reviewer: That was what I had in mind.

This differs from the `nRF Cloud MQTT API`_, which requires you to establish and maintain an `MQTT`_ connection while sending `Device Messages <nRF Cloud Device Messages_>`.

User interface
**************

> Once the device is provisioned and connected, each press of the configured button (Button 1 by default) (:kconfig:`CONFIG_REST_DEVICE_MESSAGE_BUTTON_EVT_NUM`) generates a device-to-cloud button press `Device Message <nRF Cloud Device Messages_>`_ over REST.

Reviewer: ```suggestion
Once the device is provisioned and connected, each press of the configured button (**Button 1** by default) (:kconfig:`CONFIG_REST_DEVICE_MESSAGE_BUTTON_EVT_NUM`) generates a device-to-cloud button press `Device Message <nRF Cloud Device Messages_>`_ over REST.
```

> These messages are sent to the non-bulk Device to Cloud Topic.

Reviewer: ```suggestion
These messages are sent to the non-bulk Device of a Cloud Topic.
```

See `topics used by devices running the nRF Cloud library <nRF Cloud MQTT Topics_>`_ for details.

> The configured LTE LED (LED 1 by default) (:kconfig:`CONFIG_REST_DEVICE_MESSAGE_LTE_LED_NUM`) is lit once an LTE connection is established and JWT tokens are ready to be generated.

Reviewer: ```suggestion
The configured LTE LED (**LED** 1 by default) (:kconfig:`CONFIG_REST_DEVICE_MESSAGE_LTE_LED_NUM`) is lit once an LTE connection is established and JWT tokens are ready to be generated.
```

> The configured Send LED (LED 2 by default) (:kconfig:`CONFIG_REST_DEVICE_MESSAGE_SEND_LED_NUM`) is lit when a REST request is being sent.

Reviewer: ```suggestion
The configured Send LED (**LED 2** by default) (:kconfig:`CONFIG_REST_DEVICE_MESSAGE_SEND_LED_NUM`) is lit when a REST request is being sent.
```

.. _nrf_cloud_rest_device_message_provisioning:

Provisioning
************

Your device must be provisioned for this sample to function.

> To provision your device:

Reviewer: ```suggestion
Complete the following steps to provision your device:
```

> #. Enable the :kconfig:`CONFIG_REST_DEVICE_MESSAGE_DO_JITP` option
> #. Press the button (configured with :kconfig:`CONFIG_REST_DEVICE_MESSAGE_DO_JITP`) when prompted at startup
> #. Follow the instructions printed to UART for just-in-time provisioning (JITP)

Reviewer: ```suggestion
1. Enable the :kconfig:`CONFIG_REST_DEVICE_MESSAGE_DO_JITP` option.
#. Press the button (configured with :kconfig:`CONFIG_REST_DEVICE_MESSAGE_DO_JITP`) when prompted at startup.
#. Follow the instructions for just-in-time provisioning (JITP) printed to UART.
```

CONFIG_REST_DEVICE_MESSAGE_SEND_LED_NUM - Send LED number
   This configuration option defines which LED is used to indicate a REST request is being sent

.. kconfig:: CONFIG_REST_DEVICE_MESSAGE_BUTTON_EVT_NUM

CONFIG_REST_DEVICE_MESSAGE_BUTTON_EVT_NUM - Button number
   This configuration option defines the button to use for device interactions.

.. kconfig:: CONFIG_REST_DEVICE_MESSAGE_DO_JITP

CONFIG_REST_DEVICE_MESSAGE_DO_JITP - Enable prompt to perform JITP over REST
   This configuration option defines whether the application prompts the user for just-in-time provisioning on startup.

Building and running
********************

.. |sample path| replace:: :file:`samples/nrf9160/nrf_cloud_rest_device_message`

.. include:: /includes/build_and_run_nrf9160.txt

The configuration file for this sample is located in :file:`samples/nrf9160/nrf_cloud_rest_device_message`.
See :ref:`configure_application` for information on how to configure the parameters.

Querying Device Messages over REST API
**************************************

To query the Device Messages received by the `nRF Cloud`_ backend, send a GET request to the `ListMessages <nRF Cloud REST API ListMessages_>`_ endpoint.

> 

Reviewer: Remove an extra empty line.

.. _nrf_cloud_rest_device_message:

nRF9160: nRF Cloud REST Device Message
######################################

.. contents::
   :local:
   :depth: 2

> The REST Device Message sample demonstrates how to use the `nRF Cloud REST API`_ to send a device message using the SendDeviceMessage endpoint. The d2c topic is used in non-bulk mode.

Reviewer 2: it might be useful to add a link to https://docs.nrfcloud.com/Reference/DeviceMessages/

Author: Yes, that would certainly be useful! I will do so.

> The REST Device Message sample demonstrates how to use the `nRF Cloud REST API`_ to send a `Device Message <nrf Cloud Device messages_>`_ using the SendDeviceMessage endpoint. The d2c topic is used in non-bulk mode.

Reviewer 2: not sure about this formatting with the `<` `>`?

Author: It's a custom text link:
![custom link](https://user-images.githubusercontent.com/97475097/156257224-1d36127f-f9ec-4b41-88a3-f67444a3edc6.PNG)

Requirements
************

The sample supports the following development kits:

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading
   :rows: nrf9160dk_nrf9160_ns

The sample requires an `nRF Cloud`_ account.

.. note::
   This sample requires modem firmware v1.3.x or later.

Overview
********

> Pressing the button sends a buttonpress message that can be viewed in the nRF Cloud terminal. The device must already be provisioned. However, you can provision the device over REST by enabling :kconfig:`CONFIG_REST_DEVICE_MESSAGE_DO_JITP` and following the instructions printed to UART upon boot.

Reviewer 2: `buttonpress` --> `button press` here and elsewhere.

> Pressing the button sends a button press message that can be viewed in the `nRF Cloud`_ device terminal (check the Device to Cloud channel). The device must already be provisioned. However, you can provision the device over REST by enabling :kconfig:`CONFIG_REST_DEVICE_MESSAGE_DO_JITP` and following the instructions printed to UART upon boot.

Author: I have mostly applied these changes, but in a couple areas I made different alterations than the ones you suggested. Are the alternative alterations I made still acceptable?

> Pressing the button sends a button press message that can be viewed in the `nRF Cloud`_ device terminal (check the Device to Cloud channel). The device must already be provisioned. However, you can provision the device over REST by enabling :kconfig:`CONFIG_REST_DEVICE_MESSAGE_DO_JITP` and following the instructions printed to UART upon boot.

Reviewer: Not a big fan of the construction "you can do so by ...ing". It is not wrong by any means, but just hurts my eyes.

> Pressing the button sends a button press message that can be viewed in the `nRF Cloud`_ device terminal (check the Device to Cloud channel). The device must already be provisioned. However, you can provision the device over REST by enabling :kconfig:`CONFIG_REST_DEVICE_MESSAGE_DO_JITP` and following the instructions printed to UART upon boot.

Author: How about:

`If your device is not yet provisioned, you can provision it over REST by enabling ...`

Would that read better for you?

> Pressing the button sends a button press message that can be viewed in the `nRF Cloud`_ device terminal (check the Device to Cloud channel). The device must already be provisioned. However, you can provision the device over REST by enabling :kconfig:`CONFIG_REST_DEVICE_MESSAGE_DO_JITP` and following the instructions printed to UART upon boot.

Reviewer: Not really as it now repeats provision. And you are not happy with my suggestion. Although I don't like the "do something by ...ing" construction, let's go with "... , you can do it over the REST by enabling..."

Limitations
***********

This sample requires the network carrier to provide date and time to the modem.
Without a valid date and time, the modem cannot generate JWTs with an expiration time.

User interface
**************

Once provisioned and connected, each press of the configured button (configured using the :kconfig:`CONFIG_REST_DEVICE_MESSAGE_BUTTON_EVT_NUM` option) generates a device to cloud button press `Device Message <nRF Cloud Device Messages_>`_ over REST.

If you have the option :kconfig:`CONFIG_REST_DEVICE_MESSAGE_DO_JITP` enabled and you press the button when prompted at startup, it will request just-in-time provisioning (JITP) through REST with nRF Cloud.
This is useful when initially provisioning and associating a device on nRF Cloud.
You only need to do this once for each device.

> The configured LED (:kconfig:`CONFIG_REST_DEVICE_MESSAGE_LED_NUM`) will light once an LTE connection is established and JWT tokens are ready to be generated. Assuming the device is provisioned, this is analogous to being connected to nRF Cloud, though the REST API does not truly employ connections beyond the transient connections inherent to HTTP requests.

Author: I was trying to communicate that an LTE connection and valid JWTs are the REST API's closest equivalent to being "connected" to nRF Cloud -- but only if the device has already been provisioned.

Maybe it would be easier to delete the sentence outright. Is that fine by you?

> The configured LED (:kconfig:`CONFIG_REST_DEVICE_MESSAGE_LED_NUM`) will light once an LTE connection is established and JWT tokens are ready to be generated. Assuming the device is provisioned, this is analogous to being connected to nRF Cloud, though the REST API does not truly employ connections beyond the transient connections inherent to HTTP requests.

Reviewer 3: Maybe it should be simpler, like "REST connections are transient, unlike MQTT connections which are usually long-lived."

Or alternatively, use one of the other LEDs on the 9160DK to blink on and off when a REST transfer occurs, to make this visible to the user.

> The configured LED (:kconfig:`CONFIG_REST_DEVICE_MESSAGE_LED_NUM`) will light once an LTE connection is established and JWT tokens are ready to be generated. Assuming the device is provisioned, this is analogous to being connected to nRF Cloud, though the REST API does not truly employ connections beyond the transient connections inherent to HTTP requests.

Author: I think I can reasonably do both of those things. I will actually implement blinking tomorrow, though, since i am occupied testing another sample.

Configuration
*************
|config|

Configuration options
=====================

Check and configure the following configuration options for the sample:

.. kconfig:: CONFIG_REST_DEVICE_MESSAGE_LED_NUM

CONFIG_REST_DEVICE_MESSAGE_LED_NUM - Enable LED
   This configuration option defines if the LED is to be used.

.. kconfig:: CONFIG_REST_DEVICE_MESSAGE_BUTTON_EVT_NUM

CONFIG_REST_DEVICE_MESSAGE_BUTTON_EVT_NUM - Button number
   This configuration option defines the button to use for device interactions.

.. kconfig:: CONFIG_REST_DEVICE_MESSAGE_DO_JITP

CONFIG_REST_DEVICE_MESSAGE_DO_JITP - Enable prompt to perform JITP via REST

>    This configuration option defines if the application will prompt the user for just-in-time provisioning on startup.

Reviewer: ```suggestion
   This configuration option defines if the application prompts the user for just-in-time provisioning on startup.
```

Building and running
********************

.. |sample path| replace:: :file:`samples/nrf9160/nrf_cloud_rest_device_message`

.. include:: /includes/build_and_run_nrf9160.txt

The configuration file for this sample is located in :file:`samples/nrf9160/nrf_cloud_rest_device_message`.

> See :ref:`configure_application` for information on how to configure the parameters.

Reviewer 3: Do these messages show up in the UI on nRF Cloud?

If so, is there a limitation on this? For example, the device had to have already been provisioned for MQTT access.

If not, how does the customer verify that the messages arrive at nRF Cloud?  For example, is there a CURL command they could issue to retrieve historic data?

Author: The button messages and hello world message appear on the front end, yep!

Yes, the device must already be provisioned. JITP can be requested if it was not previously provisioned

I will try to make that clearer in documentation

Reviewer 3: You can put a link from README.rst to the nRF Cloud docs, through a doc symbol defined in sdk-nrf/doc/nrf/links.txt.  There is a section of links there under the heading, where you could add one that you need:
```
... ### Source: nrfcloud.com

Author: Sure, are you wanting me to dig up some documentation on accessing device messages via CURL and include that? We already link to Device Message documentation and to nrfcloud.com

Reviewer 3: Yes, thanks.  Let me know if you run into trouble.  It's a nice-to-have, for customers that register a device for REST access but not nRF Cloud MQTT access.
